src/asyncify/asyncify_switch_impl.c

```diff
@@ -0,0 +1,286 @@
+    target_fiber = fiber;
+    // Save payload on target fiber.
+    target_fiber->arg = arg;
+    target_fiber->fiber_arg = active_fiber;
+    // Change fiber status.
+    active_fiber->state = YIELDING;
+    // Start unwinding.
+    asyncify_start_unwind(&active_fiber->stack);
+    return NULL;
+  } else {
+    // Otherwise we must be switched to
+    asyncify_stop_rewind();
+    active_fiber->state = ACTIVE;
+    fiber = active_fiber->fiber_arg; 
+    // Note(dhil): need to be a bit careful here to make sure `fiber` is correctly updated, probably needs to be a pointer-to-a-pointer to be correct.
+    return active_fiber->arg;
+  }
+}
+
+// Noop when stack pooling is disabled.
+void fiber_init(void) {
+#if defined STACK_POOL_SIZE && STACK_POOL_SIZE > 0
+  pool.next = STACK_POOL_SIZE;
+  for (uint32_t i = 0; i < STACK_POOL_SIZE; i++) {
+    stack_pool_reclaim(&pool, fiber_stack_alloc(default_stack_size));
+  }
+  assert(pool.next == 0);
+#endif
+}
+
+// Noop when stack pooling is disabled.
+void fiber_finalize(void) {
+#if defined STACK_POOL_SIZE && STACK_POOL_SIZE > 0
+  assert(pool.next == 0);
+  for (uint32_t i = 0; i < STACK_POOL_SIZE; i++) {
+    fiber_stack_free(stack_pool_next(&pool));
+  }
+  assert(pool.next == STACK_POOL_SIZE);
+#endif
+}
+
+void *fiber_main(void *(*main)(void*), void* arg) {
+
+  // Make sure the main function has been allocated a fiber
+  assert(main != NULL);
+
+  // Initialize fibers
+  fiber_init();
+
+  // Allocate the main fiber (running the initial function of our program)
+  fiber_t main_fiber = fiber_alloc(main);
+  active_fiber = main_fiber;
+  active_fiber->arg = arg;
+
+  void *result = NULL;
+
+  while (active_fiber->state == ACTIVE || target_fiber != NULL) {
+    // Check whether we need to switch to a new fiber.
+    if (target_fiber != NULL) {
+      // We need to store the caller on the target, such that we can
+      // retrieve it inside the target fiber.
+      active_fiber = target_fiber;
+    }
+
+    // If the fiber was previously yielding then we need to rewind the
+    // state.
+    if (active_fiber->state == YIELDING) {
+      asyncify_start_rewind(&active_fiber->stack);
+    }
+
+    // Run the fiber
+    result = active_fiber->entry(active_fiber->arg /* needs to have been set at the yield point */);
+    asyncify_stop_unwind();
+
+    // If the active fiber isn't yielding at this point, it should be done
+    if (active_fiber->state != YIELDING){
+      active_fiber->state = DONE;
+      // Restart the fiber we just switched from, if it's yielding
+      if (active_fiber->fiber_arg->state == YIELDING){
+        active_fiber = active_fiber->fiber_arg;
+        active_fiber->state = ACTIVE;
+        target_fiber = NULL;
+      }
+    } else {
+      // Restore the previously executing fiber.
+      active_fiber = active_fiber->fiber_arg;
+    }
+    
+    // If the main fiber is done, we're finished (I'm not too sure about this)
+    if (main_fiber->state == DONE) {
+      break;
+    }
```

Comment:
```suggestion
```
I think this will be unnecessary. As soon as one fiber is done, everything is done (unless it did a `fiber_return_switch`).